lex/lexer.ml

```diff
@@ -0,0 +1,306 @@
+    | '{' -> incr brace_depth
+    | '}' -> decr brace_depth
+    | _ -> ()
+    end;
+    incr i
+  done;
+  advance lexbuf !i;
+  if !brace_depth <> 0
+  then lex_action lexbuf
+;;
+
+let rec lex_string lexbuf =
+  let buff = get_next_buff lexbuf 200 in
+  let i = ref 0 in
+  let continue = ref true in
+  let str = ref [] in
+  while (!continue = true) && (!i < Bytes.length buff) do
+    begin match Bytes.get buff !i with
+    | '\\' -> begin match Bytes.get buff (!i+1) with
+      | '\\' -> str := '\\' :: !str
+      | '\'' -> str := '\'' :: !str
+      | '"' -> str := '"' :: !str
+      | 'n' -> str := '\010' :: !str
+      | 't' -> str := '\009' :: !str
+      | 'b' -> str := '\008' :: !str
+      | 'r' -> str := '\013' :: !str
+      | ' ' -> str := ' ' :: !str
+      (* We skip many cases that do not happen in lexer.mll, such as \\[0-9]{3},
+       * or \\ before a new line. *)
+      | _ -> raise Bad_rule
+      end; incr i
+    | '"' -> continue := false
+    | c -> str := c :: !str
+    end;
+    incr i
+  done;
+  advance lexbuf !i;
+  let buf = Buffer.create (List.length !str) in
+  List.iter (Buffer.add_char buf) !str;
+  if !continue = true
+  then (Buffer.contents buf)::(lex_string lexbuf)
+  else (Buffer.contents buf)::[];;
+
+let lex_comment lexbuf =
+  let buff = get_next_buff lexbuf 200 in
+  let star = ref false in
+  let i = ref 0 in
+  let continue = ref true in
+  while !continue do
+    begin match (Bytes.get buff !i) with
+    | '*' -> star := true
+    | ')' -> if !star then continue := false
+    | _ -> star := false
+    end;
+    incr i
+  done;
+  advance lexbuf !i
+;;
+
+let lex_ident lexbuf =
+  let buff = get_next_buff lexbuf 100 in
+  if (Bytes.length buff) = 0 then raise Bad_rule;
+  let first = Bytes.get buff 0 in
+  (* ['A'-'Z' 'a'-'z' '_' '\192'-'\214' '\216'-'\246' '\248'-'\255'] *)
+  if (('A' <= first) && (first <= 'Z')) || (('a' <= first) && (first <= 'z'))
+        || (first = '_') || (('\192' <= first) && (first <= '\214'))
+        || (('\216' <= first) && (first <= '\246'))
+        || (('\248' <= first) && (first <= '\255'))
+  then begin
+          let continue = ref true in
+          let i = ref 1 in
+          while !continue do
+            (* ['A'-'Z' 'a'-'z' '_' '\192'-'\214' '\216'-'\246' '\248'-'\255' '\'' '0'-'9'] *)
+            let chr = Bytes.get buff !i in
+            if (('A' <= chr) && (chr <= 'Z')) || (('a' <= chr) && (chr <= 'z'))
+                || (chr = '_') || (('\192' <= chr) && (chr <= '\214'))
+                || (('\216' <= chr) && (chr <= '\246'))
+                || (('\248' <= chr) && (chr <= '\255')) || (chr = '\'')
+                || (('0' <= chr) && (chr <= '9'))
+            then incr i
+            else continue := false
+          done;
+          advance lexbuf !i;
+          Bytes.to_string (Bytes.sub buff 0 !i)
+  end else raise Bad_rule
+
+let decimal_code  c d u =
+  100 * (Char.code c - 48) + 10 * (Char.code d - 48) + (Char.code u - 48)
```

Comment:
I think that `Char.code '0'` would be more readable than `48` here.

Reply:
I copied this function directly from the ocaml source code ^^